ENH: Check matching inputs/outputs in umath generation

aragilar opened the pull request:
This is a small patch which flags typos and similar problems when modifying the umath subsystem.

aragilar committed:
ENH: Check matching inputs/outputs in umath generation

eric-wieser commented on numpy/core/code_generators/generate_umath.py:
```diff
@@ -97,6 +97,8 @@ def TD(types, f=None, astype=None, in_=None, out=None, simd=None):
     elif out is None:
         out = (None,) * len(types)
+    if len(in_) != len(out):
+        raise RuntimeError("Number of inputs don't match outputs")
```
Would be better as
```
elif len(out) != len(types):
    raise ValueError(...)
```
and same for the inputs

aragilar replied:
Yeah, that makes far more sense.

eric-wieser reviewed (approved):
LGTM, waiting on tests to pass

aragilar commented:
I've updated to checking against types, and replaced an assertion above with an explicit test (which is safer in case the assertion gets optimised out).

eric-wieser commented:
Thanks @aragilar!
